### ISSUE TITLE
Enhancement: Creation of teams, groups, workspaces

### ISSUE THREAD

**Issue by dvejsada** [danny-avila/LibreChat#4730]
### What features would you like to see added?

Option to **assign** **users** **to** **teams** and have some of the settings (e g. access to endpoints and tools, setting API keys) and sharing options (e.g. uploaded documents and prompts on team bases, rather than only have option of share to none or to all.

### More details

Just to provide an example, in which scenario it may be useful. We are starting to use librechat in our office and we have various departments with different needs. For example, marketing departments needs access to dalle and endpoint that will be good at writing interesting texts and they will have some custom prompts for their uses. However, lawyers or IT will have completely different needs (IT will need endpoint best for programing) and they will have different custom prompts. If all is shared with all, it will quickly become very cluttered with tens of custom prompts and various endpoints.

### Which components are impacted by your request?

General, UI

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by owengo** [comment]
I started a discussion about "groups" https://github.com/danny-avila/LibreChat/discussions/4614
Maybe we could merge the requests to have more interest?

**Comment by danny-avila** [comment]
This is for certain a planned feature: https://github.com/danny-avila/LibreChat/discussions/4614

I'll drop what @owengo wrote here: 

---

There are users which can have different ROLES, but there is no "group" which would allow:

To share
-  prompts
- assistants
- presets
- conversations

inside a group ( or team, whatever you call it )

The groups would probably have to be a collection in the database, but it would be possible to sync it from

- ldap config ( when ldap auth is configured )
- google groups ( when google auth is configured )

This would allow administrators to manage the teams in the company's directory, and each group/team would have a level of isolation and autonomy to share between them.
Note that also some configurations such as available assistants, and even the modelSpec could be "personalized" for specific groups/teams.

**Comment by dvejsada** [comment]
> This is for certain a planned feature: https://github.com/danny-avila/LibreChat/discussions/4614
> 
> I'll drop what @owengo wrote here: 
> 
> ---
> 
> There are users which can have different ROLES, but there is no "group" which would allow:
> 
> To share
> -  prompts
> - assistants
> - presets
> - conversations
> 
> inside a group ( or team, whatever you call it )
> 
> The groups would probably have to be a collection in the database, but it would be possible to sync it from
> 
> - ldap config ( when ldap auth is configured )
> - google groups ( when google auth is configured )
> 
> This would allow administrators to manage the teams in the company's directory, and each group/team would have a level of isolation and autonomy to share between them.
> Note that also some configurations such as available assistants, and even the modelSpec could be "personalized" for specific groups/teams.
> 
> 
> 

Do you have any idea about the possible implementation timeline? We are starting deployment of Librechat in our firm tomorrow and this would be great to have before we end the limited trial period and extend to all employees. However, it is a great app anyway, thanks!

**Comment by Odrec** [comment]
Would be nice also to allow group chats with different people in a team and the llms.

**Comment by Micheletto** [comment]
Group sharing options for Agents would be great too. At the moment agents are either not shared, or shared to everyone.

**Comment by mazierovictor** [comment]
@danny-avila
That would be incredible. Is there any forecast for development?

**Comment by rubentalstra** [comment]
@mazierovictor PR 90% finished.

**Comment by ronniekolk** [comment]
Hi Ruben,

This feature — leveraging OpenID (Entra ID)  groups and roles, especially for sharing capabilities with dedicated groups for agents — would really help us a lot in many areas of our company.
Are there any updates or plans regarding this?

**Comment by danny-avila** [comment]
We have an architecture for groups and sharing agents to specific users and groups almost ready to go.

**Comment by daveAtCaptiv8** [comment]
I have a version that I have written that allows users, assistants, agents, and conversations to be shared with other members of the same team. There is a simple UI for assistants, agents, and conversations - but users still. needs to be added direct to DB. It is current with 0.7.9 and has some UI embellishment to let you know when you are sharing a chat.

Conversation sharing allows to humans to join the same conversation and participate in it. Right now it uses polling to sync. This changes the conversation from one human to many bots - to many to many - which is really nice.

Let me know if anyone wants a walkthrough - it still needs a bit of cleaning - but is entirely functional

**Comment by tomliptrot** [comment]
> We have an architecture for groups and sharing agents to specific users and groups almost ready to go.

@danny-avila has there been any progress on this? A rough time estimate would be really helpful. Thanks.

**Comment by cloudsolutiongmbh** [comment]
+1 for this feature, would be very great and unlock a lot of usecases.

**Comment by tomliptrot** [comment]
Have you set up the GraphAPI permissions? (see [here](https://www.librechat.ai/docs/configuration/authentication/OAuth2-OIDC/azure#adding-graph-api-permissions)) 

- `User.Read` - Sign in and read user profile
- `People.Read` - Read user contacts
- `GroupMember.Read.All` - Read all group memberships
- `User.ReadBasic.All` - Read all users’ basic profiles

Also, I think we had this problem (where groups didn't show up) for one user account. We deleted and recreated that account and the problem went away. We never worked out why it had happened.

I don't know anything about the second question.

**Comment by bsh3p** [comment]
Very much needed feature.

Our org has so many different user groups where we want to set the models they need.

For example, we permit certain people to use Opus but we don't want everybody using it for obvious reasons.

I don't know if it falls under this feature request, however we'd ideally want to be able to name models per user group.

For example, we might want to call Sonnet 4.5 "Advanced Requests" for most user groups but for content teams who use this and Opus, we may want to call Sonnet "General Content Creation" and Opus "Advanced Content Creation".

**Comment by danny-avila** [comment]
Closing as completed — the groups architecture landed, and this thread reads as pending when it isn't.

What's in main now:

- **Group principals for sharing.** `packages/data-schemas/src/schema/aclEntry.ts` and `systemGrant.ts` carry `principalType`, so agents, prompt groups, files and MCP servers can be shared to a *group* rather than only "just me" or "everyone" (`packages/data-provider/src/permissions.ts` covers agents, prompts, mcpServers, remoteAgents).
- **The sharing UI.** `client/src/components/Sharing/PeoplePicker/*`, gated by `interface.peoplePicker: { users, groups, roles }` in `librechat.example.yaml`.
- **Per-role and per-group settings.** The admin config stores overrides keyed by `principalType` (`packages/data-schemas/src/schema/config.ts`). That's how "marketing gets DALL·E and these endpoints, legal gets those" is expressed today — as a scoped config override rather than a separate teams config, which is why it isn't obvious from the docs.

Shipped in `949682ef` (Agent Marketplace, which added the `PEOPLE_PICKER` permission with `VIEW_USERS`/`VIEW_GROUPS`) and `81b32e40` (sharing components and permission bits).

@tomliptrot — thanks for confirming it on v0.8.0 and for posting the Graph API permission list in this thread; missing `GroupMember.Read.All` is the setup step most people hit, and your comment is the best answer here. @changkun, on your second question: gating the agent builder to advanced users is a role permission now (the `AGENTS` use/create/share permissions per role), not a code change.

**Two asks from this thread did not ship, and neither is tracked anywhere else:**

- @Odrec's **group chat** — several humans in one conversation alongside the model. Nothing in main does this; @daveAtCaptiv8's fork is still the only implementation described.
- @bsh3p's **per-group model labels** — calling Sonnet "Advanced Requests" for most groups and "General Content Creation" for the content team. Group-scoped model *access* is expressible via config overrides; per-group *renaming* is not.

Please open focused issues for either one. They'll move faster than a 24-comment thread whose headline feature is already done.

Severity: n/a, feature. Priority: was high — 46 reactions across 15+ distinct commenters; the two residuals are low on their own.
